### PR TITLE
feat(testing): add implicit test module

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

**Comment by juliemr** [comment]
Please update commit message to say that `addProviders` still works.

**Comment by juliemr** [comment]
Nevermind, that is only for the compiler.

File: modules/@angular/core/testing/test_injector.ts [hunk 1]
```diff
@@ -91,20 +182,23 @@ export function getTestInjector() {
  * @experimental
  */
 export function setBaseTestProviders(
-    platformProviders: Array<Type|Provider|any[]>,
-    applicationProviders: Array<Type|Provider|any[]>) {
+    platformCompilerProviders: Array<Type|Provider|any[]>,
+    platformProviders: Array<Type|Provider|any[]>, appModule: Type): Promise<any> {
```
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Make this synchronous again. Would only be async if we had built in `precompile` components in our default platforms, which we never will!

File: modules/@angular/core/testing/test_injector.ts [hunk 2]
```diff
@@ -91,20 +182,23 @@ export function getTestInjector() {
  * @experimental
  */
 export function setBaseTestProviders(
```
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Add docs to commit message:
* compiler providers
* how to switch, i.e. use BrowserDynamicTestModule

File: modules/@angular/core/testing/test_injector.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Keep it sync for now. We can add a special function for async later.

File: modules/@angular/core/testing/test_injector.ts [hunk 3]
```diff
@@ -146,35 +241,51 @@ export function resetBaseTestProviders() {
...
   } else {
-    // Return a synchronous test method with the injected tokens.
-    return () => { return getTestInjector().execute(tokens, fn); };
+    return () => {
+      // Return a asynchronous test method with the injected tokens.
+      if (Zone.current.name.toLowerCase().indexOf('asynctestzone') >= 0) {
+        return testInjector.createInjectorAsync().then(() => testInjector.execute(tokens, fn));
+      } else {
+        // Return a synchronous test method with the injected tokens.
+        testInjector.createInjectorSync();
```
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Add try/catch around here to give the user a better message what to do.
E.g. add an `async`  or `fakeAsync` around the test.

File: modules/@angular/core/testing/test_injector.ts [hunk 4]
```diff
@@ -146,35 +241,51 @@ export function resetBaseTestProviders() {
...
+        testInjector.execute(tokens, fn);
+        return completer.promise;
+      });
     };
   } else {
-    // Return a synchronous test method with the injected tokens.
-    return () => { return getTestInjector().execute(tokens, fn); };
+    return () => {
+      // Return a asynchronous test method with the injected tokens.
+      if (Zone.current.name.toLowerCase().indexOf('asynctestzone') >= 0) {
```
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Add TODO: better means of detecting the zone kind

File: modules/@angular/core/testing/test_injector.ts [hunk 5]
```diff
@@ -184,9 +295,25 @@ export class InjectSetupWrapper {
...
+}) {
+  return new InjectSetupWrapper(moduleDef);
+}
+
+
 // This is to ensure inject(Async) within InjectSetupWrapper doesn't call itself
 // when transpiled to Dart.
 var inject_impl = inject;
+
+class _DynamicModule {}
```
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Add a comment why this is needed.
Or call it `_NoopModule`

File: modules/@angular/core/testing/testing.ts [hunk 6]
```diff
@@ -133,6 +133,48 @@ export function addProviders(providers: Array<any>): void {
...
+  pipes?: any[],
+  precompile?: any[],
+  modules?: any[]
+}): void {
+  if (!moduleDef) return;
+  try {
+    testInjector.configureModule(moduleDef);
+  } catch (e) {
+    throw new Error(
+        'addModule can\'t be called after the injector has been already created for this test. ' +
```
Comment (tbosch) [inline modules/@angular/core/testing/testing.ts]: `configureModule`

File: modules/@angular/platform-browser-dynamic/testing.ts [hunk 7]
```diff
@@ -8,33 +8,61 @@
...
 
 import {BROWSER_APP_COMPILER_PROVIDERS} from './index';
 import {DOMTestComponentRenderer} from './testing/dom_test_component_renderer';
 
 export * from './private_export_testing'
 
 /**
+ * Default compiler providers for testing.
+ *
+ * This is experimental as we plan to use an `@AppModule` to define the testing enviroment,
```
Comment (tbosch) [inline modules/@angular/platform-browser-dynamic/testing.ts]: Remove the comment here

File: modules/@angular/platform-browser-dynamic/testing.ts [hunk 8]
```diff
@@ -8,33 +8,61 @@
...
+     useClass: MockDirectiveResolver},
+    {provide: ViewResolver,
+     useClass: MockViewResolver}
+  ]
+];
+
+/**
  * Default platform providers for testing.
  *
- * @stable
```
Comment (tbosch) [inline modules/@angular/platform-browser-dynamic/testing.ts]: Remove comment chnage

File: modules/@angular/platform-browser-dynamic/testing.ts [hunk 9]
```diff
@@ -8,33 +8,61 @@
...
 export const TEST_BROWSER_DYNAMIC_PLATFORM_PROVIDERS: Array<any /*Type | Provider | any[]*/> =
     [TEST_BROWSER_PLATFORM_PROVIDERS];
 
 /**
  * Default application providers for testing.
  *
- * @stable
+ * This is experimental as we plan to use an `@AppModule` to define the testing enviroment,
+ * instead of using arrays of providers.
+ * @experimental
```
Comment (tbosch) [inline modules/@angular/platform-browser-dynamic/testing.ts]: Make deprecated.

File: modules/@angular/platform-browser/test/testing_public_spec.ts [hunk 10]
```diff
@@ -203,6 +228,114 @@ export function main() {
...
+    describe('setting up the compiler', () => {
+      beforeEach(() => {
+        let xhrGet = jasmine.createSpy('xhrGet').and.returnValue(Promise.resolve('Hello world!'));
+        configureCompiler({providers: [{provide: XHR, useValue: {get: xhrGet}}]});
+      });
+      it('should use set up providers',
+         fakeAsync(inject([TestComponentBuilder], (tcb: TestComponentBuilder) => {
+           let compFixture = tcb.createFakeAsync(CompWithUrlTemplate);
+           expect(compFixture.nativeElement).toHaveText('Hello world!');
+         })));
```
Comment (tbosch) [inline modules/@angular/platform-browser/test/testing_public_spec.ts]: Add test for `useJit` by injecting the `CompilerConfig`

File: modules/@angular/platform-browser/testing/browser.ts [hunk 11]
```diff
@@ -54,7 +54,13 @@ export const TEST_BROWSER_PLATFORM_PROVIDERS: Array<any /*Type | Provider | any[
 /**
  * Default application providers for testing without a compiler.
  *
- * @stable
+ * This is experimental as we plan to use an `@AppModule` to define the testing enviroment,
+ * instead of using arrays of providers.
+ * @experimental
```
Comment (tbosch) [inline modules/@angular/platform-browser/testing/browser.ts]: deprecated

File: modules/@angular/platform-server/testing/server.ts [hunk 12]
```diff
@@ -35,47 +38,13 @@ export const TEST_SERVER_PLATFORM_PROVIDERS: Array<any /*Type | Provider | any[]
...
-      /* @ts2dart_Provider */ {provide: APP_ID, useValue: 'a'},
-      /* @ts2dart_Provider */ {provide: SharedStylesHost, useExisting: DomSharedStylesHost},
-      DomSharedStylesHost, ELEMENT_PROBE_PROVIDERS,
-      {provide: TestComponentBuilder, useClass: OverridingTestComponentBuilder},
-      /* @ts2dart_Provider */ {provide: DirectiveResolver, useClass: MockDirectiveResolver},
-      /* @ts2dart_Provider */ {provide: ViewResolver, useClass: MockViewResolver},
-      /* @ts2dart_Provider */ {provide: TestComponentRenderer, useClass: DOMTestComponentRenderer},
-      /* @ts2dart_Provider */ {provide: NgZone, useFactory: createNgZone}
-    ];
+export const TEST_SERVER_APPLICATION_PROVIDERS = TEST_BROWSER_DYNAMIC_APPLICATION_PROVIDERS;
```
Comment (tbosch) [inline modules/@angular/platform-server/testing/server.ts]: Deprecated

File: modules/@angular/router/karma-test-shim.js [hunk 13]
```diff
@@ -72,8 +72,10 @@ Promise.all([
   var testing = providers[0];
   var testingBrowser = providers[1];
 
-  testing.setBaseTestProviders(testingBrowser.TEST_BROWSER_DYNAMIC_PLATFORM_PROVIDERS,
-    testingBrowser.TEST_BROWSER_DYNAMIC_APPLICATION_PROVIDERS);
+  return testing.setBaseTestProviders(
```
Comment (tbosch) [inline modules/@angular/router/karma-test-shim.js]: revert this

File: test-main.js [hunk 14]
```diff
@@ -75,9 +75,10 @@ System.import('@angular/core/testing')
   .then(function(coreTesting){
     return System.import('@angular/platform-browser-dynamic/testing')
       .then(function(browserTesting){
-         coreTesting.setBaseTestProviders(
+         return coreTesting.setBaseTestProviders(
```
Comment (tbosch) [inline test-main.js]: revert promise

File: tools/cjs-jasmine/index.ts [hunk 15]
```diff
@@ -66,11 +66,12 @@ jrunner.configureDefaultReporter({showColors: process.argv.indexOf('--no-color')
 jrunner.onComplete(function(passed: boolean) { process.exit(passed ? 0 : 1); });
 jrunner.projectBaseDir = path.resolve(__dirname, '../../');
 jrunner.specDir = '';
-require('./test-cjs-main.js');
-require('zone.js/dist/jasmine-patch.js');
-distAllRequire('@angular/platform-server/src/parse5_adapter.js').Parse5DomAdapter.makeCurrent();
-specFiles.forEach((file: string) => {
-  var r = distAllRequire(file);
-  if (r.main) r.main();
+require('./test-cjs-main.js').then(() => {
```
Comment (tbosch) [inline tools/cjs-jasmine/index.ts]: remove `then`

File: tools/cjs-jasmine/test-cjs-main.ts [hunk 16]
```diff
@@ -1,6 +1,6 @@
 var testingPlatformServer = require('../../all/@angular/platform-server/testing/server.js');
 var testing = require('../../all/@angular/core/testing');
 
-testing.setBaseTestProviders(
-    testingPlatformServer.TEST_SERVER_PLATFORM_PROVIDERS,
-    testingPlatformServer.TEST_SERVER_APPLICATION_PROVIDERS);
+module.exports = testing.setBaseTestProviders(
```
Comment (tbosch) [inline tools/cjs-jasmine/test-cjs-main.ts]: remove export

File: modules/@angular/core/testing/test_injector.ts [hunk 17]
```diff
@@ -6,61 +6,152 @@
...
 import {lockRunMode} from '../src/application_ref';
 import {ListWrapper} from '../src/facade/collection';
 import {BaseException} from '../src/facade/exceptions';
 import {FunctionWrapper, isPresent} from '../src/facade/lang';
+
 import {AsyncTestCompleter} from './async_test_completer';
 
+const UNDEFINED = new Object();
+
+export const COMPILER_INJECTOR = new OpaqueToken('CompilerInjector');
```
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Make this private

File: modules/@angular/core/testing/testing.ts [hunk 18]
```diff
@@ -133,6 +133,48 @@ export function addProviders(providers: Array<any>): void {
...
+        'This is most likely because you\'ve already used the injector to inject a beforeEach or the ' +
+        'current `it` function.');
+  }
+}
+
+/**
+ * Allows overriding default compiler providers and settings
+ * which are defined in test_injector.js
+ *
+ * @stable
```
Comment (tbosch) [inline modules/@angular/core/testing/testing.ts]: Mark as experimental

**Comment by juliemr** [comment]
cc @vikerman for some changes to testing API.

File: modules/@angular/core/testing/test_injector.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Done

File: modules/@angular/core/testing/test_injector.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Done

File: modules/@angular/core/testing/test_injector.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Done

File: modules/@angular/core/testing/test_injector.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Done

File: modules/@angular/core/testing/test_injector.ts (hunk 5 above)
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Done

File: modules/@angular/core/testing/testing.ts (hunk 6 above)
Comment (tbosch) [inline modules/@angular/core/testing/testing.ts]: Done

File: modules/@angular/platform-browser-dynamic/testing.ts (hunk 7 above)
Comment (tbosch) [inline modules/@angular/platform-browser-dynamic/testing.ts]: Done

File: modules/@angular/platform-browser-dynamic/testing.ts (hunk 8 above)
Comment (tbosch) [inline modules/@angular/platform-browser-dynamic/testing.ts]: Done

File: modules/@angular/platform-browser-dynamic/testing.ts (hunk 9 above)
Comment (tbosch) [inline modules/@angular/platform-browser-dynamic/testing.ts]: Done

File: modules/@angular/platform-browser/test/testing_public_spec.ts (hunk 10 above)
Comment (tbosch) [inline modules/@angular/platform-browser/test/testing_public_spec.ts]: Done

File: modules/@angular/platform-browser/testing/browser.ts (hunk 11 above)
Comment (tbosch) [inline modules/@angular/platform-browser/testing/browser.ts]: Done, deleted.

File: modules/@angular/platform-server/testing/server.ts (hunk 12 above)
Comment (tbosch) [inline modules/@angular/platform-server/testing/server.ts]: Done, deleted.

File: modules/@angular/router/karma-test-shim.js (hunk 13 above)
Comment (tbosch) [inline modules/@angular/router/karma-test-shim.js]: Done

File: tools/cjs-jasmine/index.ts (hunk 15 above)
Comment (tbosch) [inline tools/cjs-jasmine/index.ts]: Done

File: tools/cjs-jasmine/test-cjs-main.ts (hunk 16 above)
Comment (tbosch) [inline tools/cjs-jasmine/test-cjs-main.ts]: Done

File: modules/@angular/core/testing/test_injector.ts (hunk 17 above)
Comment (tbosch) [inline modules/@angular/core/testing/test_injector.ts]: Marked as experimental for now.

File: modules/@angular/core/testing/testing.ts (hunk 18 above)
Comment (tbosch) [inline modules/@angular/core/testing/testing.ts]: Keep stable as the method `createCompiler` in `platform-browser-dynamic` is also stable.

File: tools/public_api_guard/core/testing.d.ts [hunk 19]
```diff
@@ -107,20 +131,44 @@ export declare class TestComponentRenderer {
...
 
 /** @experimental */
-export declare class TestInjector {
-    applicationProviders: Array<Type | Provider | any[] | any>;
-    platformProviders: Array<Type | Provider | any[] | any>;
-    addProviders(providers: Array<Type | Provider | any[] | any>): void;
-    createInjector(): ReflectiveInjector;
+export declare class TestInjector implements Injector {
+    appModule: Type;
+    compilerFactory: (config: {
```
Comment (juliemr) [inline tools/public_api_guard/core/testing.d.ts]: Please add a public typing for the compiler factory, as discussed.

File: modules/@angular/platform-browser-dynamic/testing.ts [hunk 20]
```diff
@@ -6,35 +6,61 @@
...
+
 /**
- * Default platform providers for testing.
+ * Creates a compiler for testing
  *
  * @stable
  */
-export const TEST_BROWSER_DYNAMIC_PLATFORM_PROVIDERS: Array<any /*Type | Provider | any[]*/> =
-    [TEST_BROWSER_PLATFORM_PROVIDERS];
+export function browserTestCompiler(
```
Comment (juliemr) [inline modules/@angular/platform-browser-dynamic/testing.ts]: Maybe this should be called `browserTestCompilerFactory` or another name that corresponds to the public typing for compiler factory.

Comment (tbosch) [inline modules/@angular/platform-browser-dynamic/testing.ts]: Would like to keep it with `browserTestCompiler` as the method that we use to create the compiler for the application is `browserCompiler`, i.e. it does not have a `Factory` suffix. See https://github.com/angular/angular/blob/master/modules/%40angular/platform-browser-dynamic/index.ts#L57

**Comment by juliemr** [comment]
Please add `import` statements to the snippets in the BREAKING CHANGE documentation, so we don't have a mysterious 'testingBrowser` variable in there. Could you make it have `before` and `after` sections as well for the `setBaseTestProviders` part?

**Comment by juliemr** [comment]
As discussed offline, let's change `setBaseTestProviders` to `initTestEnvironment`

**Pull request closed without merging**